### PR TITLE
Fix order of operations in ranges to avoid overflow

### PR DESCRIPTION
**Pull request by bkamins**
A minor proposal changing the order of operations in calculation of ranges. In the old code the problem arises when operations on `stop` and `step` overflows arithmetic. In general it can occur for rational types and `Char`.

### CODE REVIEW

File: base/range.jl [hunk 1]
```diff
@@ -41,7 +41,7 @@ Range operator. `a:b` constructs a range from `a` to `b` with a step size of 1 (
 # without the second method above, the first method above is ambiguous with
 # (:)(start::A, step, stop::C) where {A<:Real,C<:Real}
 function _colon(start::T, step, stop::T) where T
-    T′ = typeof(start+step)
+    T′ = typeof(start+zero(step))
```
Comment (oxinabox) [inline base/range.jl]:
```
function _colon(start::T, step::TS, stop::T) where {T,TS}
    T′ = promote_type(T,TS)
...
```

Be more sensible?

Though, if one is defining addition, to not use type-promotion,  my suggestion wouldn't work.
But that is a pathological case, and at that point are we even sure `zero` will be defined?

Comment (bkamins) [reply base/range.jl]: I have thought about it. The problem is that when you take `'a':1:'c'` then:
```
julia> promote_type(Char, Int)
Any
```
so I have thought that assuming `TS` defines `zero` is more sensible (typical steps like numbers or `Day` or `Millisecond` define `zero`). Not 100% sure though.

Comment (oxinabox) [reply base/range.jl]: Ah, fair enough.

**Comment by JeffBezanson** [comment]
LGTM.

**Comment by bkamins** [comment]
The CI error seems unrelated.

**Pull request merged**